File: References/Curved Spaces/Source code/Shared/Geometry Games Classes - Mac-iOS/GeometryGamesModel.m
//	GeometryGamesModel.m
//
//	© 2016 by Jeff Weeks
//	See TermsOfUse.txt

#import "GeometryGamesModel.h"
#import "GeometryGamesUtilities-Common.h"

//	On MacOS,
//		stdlib.h includes Availability.h
//		which includes AvailabilityInternal.h
//		which defines __MAC_OS_X_VERSION_MIN_REQUIRED
//		which is the preferred way to check whether
//			the code is compiling on MacOS.
#include <stdlib.h>
#ifdef __MAC_OS_X_VERSION_MIN_REQUIRED
#define REPORT_LAST_MODEL_CLOSED
#endif

#ifdef REPORT_LAST_MODEL_CLOSED
//	Normally when a user quits a Mac application
//	the system terminates it as quickly as possible,
//	without explicitly deallocating each block of memory.
//	However, if we want to confirm that there were no leaks
//	in the platform-independent C code, we must instead
//	close each window to free its ModelData, and then
//	once all the ModelData objects have been freed,
//	tell the app delegate that it's time to check for leaks
//	and terminate the application.
//
//		Caution:  If a window wants to replace its model, it must
//		of course create the new one before freeing the old one,
//		because otherwise the app would terminate!  Fortunately
//		creating the new model before freeing the old one
//		is standard practice anyhow, so that in case of error
//		while creating the new model, the old model may be kept.
//
static unsigned int	gNumActiveModels = 0;
#endif


@implementation GeometryGamesModel
{
	//	Mac OS
	//		The main thread and the CVDisplayLink thread both need 
	//		to access the ModelData, so use an NSLock to prevent conflicts.
	//
	//	iOS
	//		The CADisplayLink runs on the main thread, so in principle 
	//		no lock is needed.  Nevertheless, it seems worth using
	//		locking on iOS anyhow, to keep the Mac and iOS code 
	//		as similar as possible.
	//
	//	Only the -lockModelData method may directly access
	//	the instance variable itsModelDataPROTECTED.
	//	All other code should instead check out a pointer, use it, 
	//	and then check it back in again as soon as possible:
	//
	//		ModelData	md = NULL;
	//
	//		[theModel lockModelData:&md];
	//		...
	//		use md as desired
	//		...
	//		[theModel unlockModelData:&md];
	//
	ModelData	*itsModelDataPROTECTED;
	NSLock		*itsModelDataLock;
}


- (id)init
{
	self = [super init];
	if (self != nil)
	{
		//	Let the app-specific code allocate the ModelData.
		itsModelDataPROTECTED = (ModelData *)GET_MEMORY(SizeOfModelData());
		if (itsModelDataPROTECTED == NULL)
		{
			self = nil;	//	might not be necessary to set self = nil
			return nil;
		}
		
		//	Initialize the ModelData.
		SetUpModelData(itsModelDataPROTECTED);

		//	Create a lock to ensure that the main thread and 
		//	the CVDisplayLink never access the ModelData simultaneously.
		itsModelDataLock = [[NSLock alloc] init];

#ifdef REPORT_LAST_MODEL_CLOSED
		gNumActiveModels++;
#endif
	}
	return self;
}

- (void)dealloc
{
	//	If we're getting dealloc'd, that should mean that
	//	no objects hold a pointer to us, and no objects
	//	hold itsModelDataLock.
	
	ShutDownModelData(itsModelDataPROTECTED);
	FREE_MEMORY_SAFELY(itsModelDataPROTECTED);

#ifdef REPORT_LAST_MODEL_CLOSED

	gNumActiveModels--;

#pragma clang diagnostic push
#pragma clang diagnostic ignored "-Wundeclared-selector"
	if (gNumActiveModels == 0)
		[NSApp sendAction:@selector(lastModelDidDeallocate:) to:nil from:self];
#pragma clang diagnostic pop

#endif	//	REPORT_LAST_MODEL_CLOSED
}


- (void)lockModelData:(ModelData **)aModelDataPtr
{
	GEOMETRY_GAMES_ASSERT(aModelDataPtr != NULL && *aModelDataPtr == NULL,
		"Bad input received in GeometryGamesModel -lockModelData");

	[itsModelDataLock lock];
	
	*aModelDataPtr = itsModelDataPROTECTED;
}

- (void)unlockModelData:(ModelData **)aModelDataPtr
{
	GEOMETRY_GAMES_ASSERT(aModelDataPtr != NULL && *aModelDataPtr != NULL,
		"Bad input received in GeometryGamesModel -unlockModelData");

	[itsModelDataLock unlock];
	
	*aModelDataPtr = NULL;
}


@end

File: References/Curved Spaces/Source code/Shared/GeometryGamesClasses-Android/src/org/geometrygames/geometrygamesshared/GeometryGamesModel.java
package org.geometrygames.geometrygamesshared;


import java.util.concurrent.locks.ReentrantLock;


public class GeometryGamesModel
{
	//	The main thread and the GLSurfaceView thread both need 
	//	to access the ModelData, so use a Lock to prevent conflicts.
	//
	//	Only the lockModelData() method may directly access
	//	the instance variable itsModelDataPRIVATE.
	//	All other code should instead check out a pointer, use it, 
	//	and then check it back in again as soon as possible:
	//
	//		long	md = 0;
	//
	//		md = theModel.lockModelData();
	//		Pass md to the C code to use as desired.
	//		md = theModel.unlockModelData();
	//
	//	or to protect against Java exceptions
	//
	//		long	md = 0;
	//
	//		md = theModel.lockModelData();
	//		try
	//		{
	//			Pass md to the C code to use as desired.
	//		}
	//		finally
	//		{
	//			md = theModel.unlockModelData();
	//		}
	//
	//
	private long			itsModelDataPRIVATE;	//	C pointer to app-specific ModelData; always 64-bit
	private ReentrantLock	itsModelDataLock;

	public GeometryGamesModel()
	{
		super();
		
		//	Set up an app-specific model.
		itsModelDataPRIVATE = GeometryGamesJNIWrapper.alloc_model_data();
		GeometryGamesJNIWrapper.set_up_model_data(itsModelDataPRIVATE);

		//	Set up a lock.
		itsModelDataLock = new ReentrantLock();
	}
	
	@Override
	public void finalize() throws Throwable
	{
		//	Note:  Releasing memory allocated in C code is tricky in Java.
		//	Various articles online warn that there's no way
		//	to know when -- or even if -- finalize() will get called.
		//	We could write a custom method to release itsModelDataPRIVATE,
		//	but we'd need to manually keep track of when this ModelData
		//	was eligible to be finalized.
		
		//	Shut down the app-specific model.
		GeometryGamesJNIWrapper.shut_down_model_data(itsModelDataPRIVATE);
		itsModelDataPRIVATE = GeometryGamesJNIWrapper.free_model_data(itsModelDataPRIVATE);

		//	The Java runtime will automatically release the lock.

		super.finalize();
	}
	
	public long lockModelData()
	{
		//	lock() blocks until the lock becomes available.
		itsModelDataLock.lock();

		//	Return the ModelData pointer to the caller as a long int.
		return itsModelDataPRIVATE;
	}
	
	public long unlockModelData()
	{
		//	Unlock the lock.
		itsModelDataLock.unlock();
		
		//	Return 0 to the caller as a convenience value,
		//	so the caller may zero out its copy of the Model Data "pointer".
		return 0;
	}
}
